virttest/qemu_vm.py

```diff
@@ -2460,25 +2452,29 @@ def sort_key(dev):
             attr_info = [None, params["keyboard_layout"], None]
             add_qemu_option(devices, "k", [attr_info])
 
-        # Add disable_legacy and disable_modern options
-        virtio_devices = filter(lambda x: x.get_param('driver', '').startswith('virtio-'), devices)
+        # Add options for all virtio devices
+        virtio_devices = filter(lambda x: re.search(r"(?:^virtio-)|(?:^vhost-)",
+                                                    x.get_param('driver', '')),
+                                devices)
         for device in virtio_devices:
             dev_type = device.get_param("driver")
             # Currently virtio1.0 behaviour on latest RHEL.7.2/RHEL.7.3
             # qemu versions is default, we don't need to specify the
             # disable-legacy and disable-modern options explicitly.
-            disable_legacy = params.get("virtio_dev_disable_legacy", None)
-            disable_modern = params.get("virtio_dev_disable_modern", None)
-            iommu_platform = params.get("virtio_dev_iommu_platform", None)
-            ats = params.get("virtio_dev_ats", None)
-            if disable_legacy:
-                add_virtio_option("disable-legacy", disable_legacy, devices, device, dev_type)
-            if disable_modern:
-                add_virtio_option("disable-modern", disable_modern, devices, device, dev_type)
-            if iommu_platform:
-                add_virtio_option("iommu_platform", iommu_platform, devices, device, dev_type)
-            if ats:
-                add_virtio_option("ats", ats, devices, device, dev_type)
+            dev_info = devices.execute_qemu("-device %s,help" % dev_type)
+            dev_properties = re.findall(r"([a-zA-Z0-9_-]+)=\S+", dev_info)
+            disable_legacy = params.get("virtio_dev_disable_legacy")
+            disable_modern = params.get("virtio_dev_disable_modern")
+            iommu_platform = params.get("virtio_dev_iommu_platform")
+            ats = params.get("virtio_dev_ats")
+            if "disable-legacy" in dev_properties:
+                device.set_param("disable-legacy", disable_legacy)
+            if "disable-modern" in dev_properties:
+                device.set_param("disable-legacy", disable_modern)
+            if "iommu_platform" in dev_properties:
+                device.set_param("disable-legacy", iommu_platform)
```

Comment:
And `"iommu_platform"`? Seems you wrongly copy `"disable-legacy"` for all of them?

Reply:
+1